impl/src/planning.rs

```diff
@@ -493,6 +479,52 @@ impl<'planner> WorkspaceSubplanner<'planner> {
       })
       .collect()
   }
+
+  fn populate_sha256_index(&self) -> Result<HashMap<(String, Version), String>> {
+    // Gather the checksums for all packages in the lockfile
+    // which have them.
+    //
+    // Store the representation of the package as a tuple
+    // of (name, version) -> checksum.
+    let mut package_to_checksum = HashMap::new();
+    if let Some(lock_path) = self.files.lock_path_opt.as_ref() {
+      let lockfile = Lockfile::load(lock_path.as_path())?;
+      for package in lockfile.packages {
+        if let Some(checksum) = package.checksum {
+          package_to_checksum.insert(
+            (package.name.to_string(), package.version),
+            checksum.to_string(),
+          );
+        }
+      }
+    }
+
+    for package in &self.metadata.packages {
+      let key = (package.name.to_string(), package.version.clone());
+      if !package_to_checksum.contains_key(&key) {
+        if let Some(source) = package.source.as_ref() {
+          let source = format!("{}", source);
+          if let Some(registry_url) = source.strip_prefix("registry+") {
+            if let Ok(bare_index) = crates_index::BareIndex::from_url(&registry_url) {
+              if let Ok(index) = bare_index.open_or_clone() {
+                if let Some(crate_) = index.crate_(&package.name) {
+                  let want_version = package.version.to_string();
+                  for version in crate_.versions() {
+                    if version.version() == &want_version {
+                      package_to_checksum.insert(key, hex::encode(version.checksum()));
+                      break;
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
```

Comment:
This construction is a bit messy. Linear logic should not indent endlessly.

You can resolve most of this via early returns or continues and get back to a linear structure

```rust
    let key = (package.name.to_string(), package.version.clone());
    for package in &self.metadata.packages {
      if package_to_checksum.contains_key(&key) {
        // Checksum already populated from lockfile.
        continue;
      }
      let deprefixed_registry_src = package.source.as_ref()
        .map(|src| format!("{}", src)
        .and_then(source.strip_prefix("registry+"));
      if (deprefixed_registry_src.is_none()) {
        // Can't fetch checksum from crate which isn't from the registry
        continue;
      }
      let index = crates_index::BareIndex::from_url(&deprefixed_registry_src.unwrap())
        .and_then(|bare_index| bare_index.open_or_clone());
      if let Err(index_err) = index {
        // Could not open crate index. Maybe should log?
        continue;
      }      
      if let Some(crate_) = index.crate_(&package.name) {
        let want_version = package.version.to_string();
        for version in crate_.versions() {
           if version.version() == &want_version {
              package_to_checksum.insert(key, hex::encode(version.checksum()));
              break;
           }
        }
      }
    }

```

Reply:
(this is largely optional, though I feel a bit strongly about it, so I'd hope you feel strongly if you don't want to make the change)